### PR TITLE
customizable authorization error messages

### PR DESCRIPTION
**Pull request by gha-zund**
- introduce AuthorizationErrorMessageBuilder
- move code for generating message from AuthorizationError to AuthorizationErrorMessageBuilder
- make methods for building the message virtual, so it becomes customizable
- inject AuthorizationErrorMessageBuilder to AuthorizationValidationRule

### CODE REVIEW

File: src/Authorization.AspNetCore/AuthorizationError.cs [hunk 1]
```diff
@@ -16,8 +13,8 @@ public class AuthorizationError : ValidationError
         /// <summary>
         /// Initializes a new instance of the <see cref="AuthorizationError"/> class for a specified authorization result.
         /// </summary>
-        public AuthorizationError(INode node, ValidationContext context, OperationType? operationType, AuthorizationResult result)
-            : this(node, context, GenerateMessage(operationType, result), result)
+        public AuthorizationError(INode node, ValidationContext context, OperationType? operationType, string message, AuthorizationResult result)
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/AuthorizationError.cs:16]: Let's merge two ctors into one, move `OperationType? operationType` at the end: `OperationType? operationType = null`.

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+using System.Text;
+using GraphQL.Language.AST;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Authorization.Infrastructure;
+
+namespace GraphQL.Server.Authorization.AspNetCore
+{
+    public class AuthorizationErrorMessageBuilder : IAuthorizationErrorMessageBuilder
+    {
+        public string GenerateMessage(OperationType? operationType, AuthorizationFailure failure)
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:13]:
```suggestion
        public string GenerateMessage(OperationType? operationType, AuthorizationResult result)
```
Also `AuthorizationResult.Failure` is nullable.

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+            foreach (var requirement in failure.FailedRequirements)
+            {
+                AppendFailureLine(error, requirement);
+            }
+
+            return error.ToString();
+        }
+
+        protected virtual string GetOperationType(OperationType? operationType)
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:26]: virtual?

Comment (gha-zund) [reply src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:26]: yes. maybe someone wants to override this with other text or other language.

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs (hunk 2 above)
Comment (gha-zund) [reply src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:13]: why passing the whole AuthorizationResult?
GenearteMessage should only be called, if the result is not Suceeded. In that case, we can rely that failure is not null.
That's why failure is not nullable here.

File: src/Authorization.AspNetCore/IAuthorizationErrorMessageBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+using Microsoft.AspNetCore.Authorization;
+
+namespace GraphQL.Server.Authorization.AspNetCore;
+
+public interface IAuthorizationErrorMessageBuilder
+{
+    string GenerateMessage(OperationType? operationType, AuthorizationFailure failure);
+
+    /// <summary>
+    /// Appends the error message header for this <see cref="AuthorizationError"/> to the provided <see cref="StringBuilder"/>.
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/IAuthorizationErrorMessageBuilder.cs:14]: >  for this \<see cref="AuthorizationError"/>

What this?

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            return operationType switch
+            {
+                OperationType.Query => "query",
+                OperationType.Mutation => "mutation",
+                OperationType.Subscription => "subscription",
+                _ => "operation",
+            };
+        }
+
+        /// <inheritdoc cref="IAuthorizationErrorMessageBuilder"/>
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:37]:
```suggestion
        /// <inheritdoc/>
```

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                    break;
+
+                case OperationAuthorizationRequirement operationAuthorizationRequirement:
+                    error.Append("Required operation '");
+                    error.Append(operationAuthorizationRequirement.Name);
+                    error.Append("' was not present.");
+                    break;
+
+                case RolesAuthorizationRequirement rolesAuthorizationRequirement:
+                    if (!rolesAuthorizationRequirement.AllowedRoles.Any())
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:85]:
```suggestion
                    if (rolesAuthorizationRequirement.AllowedRoles == null || !rolesAuthorizationRequirement.AllowedRoles.Any())
```

Why did you change this line?

Comment (gha-zund) [reply src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:85]: ClaimAuthorizationRequirement.AllowedValues can be null, at least when using .NET Core 3.1. Resharper gave me a warning about that.

Comment (gha-zund) [reply src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:85]: more correct would be: in .NET Core AllowedValues is not defined as not nullable.

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs (hunk 3 above)
Comment (sungam3r) [reply src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:26]: Then overriding `AppendFailureHeader` is enough.

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs (hunk 2 above)
Comment (sungam3r) [reply src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:13]: In order not to limit the context, for example, if one inherits from `AuthorizationResult`.

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs (hunk 6 above)
Comment (sungam3r) [reply src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:85]: > ClaimAuthorizationRequirement.AllowedValues can be null

So... why did you remove check that guards from NRE?
изображение

File: src/Authorization.AspNetCore/AuthorizationValidationRule.cs [hunk 7]
```diff
@@ -17,19 +19,25 @@ public class AuthorizationValidationRule : IValidationRule
...
         private readonly IClaimsPrincipalAccessor _claimsPrincipalAccessor;
+        private readonly IAuthorizationErrorMessageBuilder _messageBuilder;
 
         /// <summary>
         /// Creates an instance of <see cref="AuthorizationValidationRule"/>.
         /// </summary>
         /// <param name="authorizationService"> ASP.NET Core <see cref="IAuthorizationService"/> to authorize against. </param>
         /// <param name="claimsPrincipalAccessor"> The <see cref="IClaimsPrincipalAccessor"/> which provides the <see cref="ClaimsPrincipal"/> for authorization. </param>
-        public AuthorizationValidationRule(IAuthorizationService authorizationService, IClaimsPrincipalAccessor claimsPrincipalAccessor)
+        /// <param name="messageBuilder">The <see cref="IAuthorizationErrorMessageBuilder"/> which is used to generate the message of an <see cref="AuthorizationError"/>. </param>
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/AuthorizationValidationRule.cs:29]:
```suggestion
        /// <param name="messageBuilder">The <see cref="IAuthorizationErrorMessageBuilder"/> which is used to generate the message for an <see cref="AuthorizationError"/>. </param>
```

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+using System.Linq;
+using System.Text;
+using GraphQL.Language.AST;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Authorization.Infrastructure;
+
+namespace GraphQL.Server.Authorization.AspNetCore
+{
+    public class AuthorizationErrorMessageBuilder : IAuthorizationErrorMessageBuilder
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs:11]: AuthorizationErrorMessageBuilder  -> DefaultAuthorizationErrorMessageBuilder

File: src/Authorization.AspNetCore/IAuthorizationErrorMessageBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+using System.Text;
+using GraphQL.Language.AST;
+using Microsoft.AspNetCore.Authorization;
+
+namespace GraphQL.Server.Authorization.AspNetCore;
+
+public interface IAuthorizationErrorMessageBuilder
+{
+    string GenerateMessage(OperationType? operationType, AuthorizationResult result);
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/IAuthorizationErrorMessageBuilder.cs:11]: Please add xml comments for all new APIs.

File: src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs (hunk 6 above)
Comment (gha-zund) [reply src/Authorization.AspNetCore/AuthorizationErrorMessageBuilder.cs]: so the warning is wrong. I didn't know that Any behaves this way. Thank you!

File: src/Authorization.AspNetCore/DefaultAuthorizationErrorMessageBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                    break;
+
+                case OperationAuthorizationRequirement operationAuthorizationRequirement:
+                    error.Append("Required operation '");
+                    error.Append(operationAuthorizationRequirement.Name);
+                    error.Append("' was not present.");
+                    break;
+
+                case RolesAuthorizationRequirement rolesAuthorizationRequirement:
+                    if (!rolesAuthorizationRequirement.AllowedRoles.Any())
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/DefaultAuthorizationErrorMessageBuilder.cs:92]: Again?

File: src/Authorization.AspNetCore/IAuthorizationErrorMessageBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+namespace GraphQL.Server.Authorization.AspNetCore;
+
+public interface IAuthorizationErrorMessageBuilder
+{
+    /// <summary>
+    /// Generates an authorization error message based on the provided <see cref="AuthorizationResult"/>
+    /// </summary>
+    /// <param name="operationType">The GraphQL operation type.</param>
+    /// <param name="result">The <see cref="AuthorizationResult"/> which is used to generate the message.</param>
+    /// <returns></returns>
```
Comment (sungam3r) [inline src/Authorization.AspNetCore/IAuthorizationErrorMessageBuilder.cs:16]:
```suggestion
```

File: src/Authorization.AspNetCore/DefaultAuthorizationErrorMessageBuilder.cs (hunk 10 above)
Comment (sungam3r) [reply src/Authorization.AspNetCore/DefaultAuthorizationErrorMessageBuilder.cs:92]: I explained that check for null is needed to avoid NRE.

File: src/Authorization.AspNetCore/IAuthorizationErrorMessageBuilder.cs (hunk 11 above)
Comment (gha-zund) [reply src/Authorization.AspNetCore/IAuthorizationErrorMessageBuilder.cs:16]: ahhhh if you do it fast, you do it twice... sorry

File: src/Authorization.AspNetCore/DefaultAuthorizationErrorMessageBuilder.cs (hunk 10 above)
Comment (sungam3r) [reply src/Authorization.AspNetCore/DefaultAuthorizationErrorMessageBuilder.cs]: > because in it's constructor an exception is thrown if attempting to create a RolesAuthorizationRequirement with AllowedRoles null

Did you need to decompile the source code to check this? In any case this is the current implementation that may vary, so it is not wise to remove the check.

**Review by Shane32 — APPROVED** [review]
